### ISSUE TITLE
Enhancement: allow nodeos to run in "no blocks.log" mode

### ISSUE THREAD

**Issue by matthewdarwin** [eosnetworkfoundation/mandel#92]
Add a new configuration option to allow nodeos to not generate any blocks.log file on disk.  Good for block producer nodes.

**Comment by linh2931** [comment]
Hi @matthewdarwin, does option name `not-generate-blocks-log` sounds good? When this option is configured, I will still allow to start from a blocks.log, just not write anything into the file.

Can you also create a separate issue for your additional comment (reversible blocks) for better tracking and easier testing?

**Comment by matthewdarwin** [comment]
Seems messy to have a blocks.log that is not kept up-to-date.

**Comment by linh2931** [comment]
I am not aware of reversible blocks issue. Do you have an issue number?

**Comment by heifner** [comment]
Yes, #180 is in `3.1.0-rc1`

**Comment by linh2931** [comment]
> Seems messy to have a blocks.log that is not kept up-to-date.

OK. When `not-generate-blocks-log` is configured, `blocks.log` file will not opened and replay from it will not allowed.

**Comment by spoonincode** [comment]
Another possibility for enabling "no blocks.log mode" would be via `block-log-retain-blocks = 0` (currently the setting is required to be 1 or more)

Of course, as is, that option's behavior is to prune immediately which is at odds of
> When this option is configured, I will still allow to start from a blocks.log, just not write anything into the file.

There is ongoing discussion that the `block-log-retain-blocks` option should not prune until replay is complete; maybe that aligned better with the above quote that would allow starting  with a blocks.log still (and then remove the blocks.log once replay is complete?)

**Comment by linh2931** [comment]
I think `not-generate-blocks-log` conveys the intention of the nodeos operator better than `block-log-retain-blocks` being 0. It also avoids the call to block_log's `append`, which saves it the trouble from deciding not to write to the blocks.log if `block-log-retain-blocks` is 0.

**Comment by spoonincode** [comment]
I'm not suggesting leaning on the existing pruning code to support no blocks.log mode; as you mention the logic should be different (i.e. no file, no `append()` calls, etc). I'm suggesting to at least consider overloading the option instead of adding a new option. After all, someone who wants no block log wants to _retain zero blocks_.

**Comment by linh2931** [comment]
Thanks Matt for clarification. It is clear to me now. I agree to use `block-log-retain-blocks = 0`, so I don't add one more new option to the code and the users.

**Comment by matthewdarwin** [comment]
If `block-log-retain-blocks = 0` I would expect blocks.log file would be empty (or not there).  Not sure how replay is handled.  Maybe remove the file after blocks.log is done replay?

**Comment by linh2931** [comment]
We can remove blocks.log after a replay. A warning would be added to the description of `block-log-retain-blocks = 0` behavior.

**Comment by linh2931** [comment]
In chain_plugin.cpp, `block-log-retain-blocks` is only available if `cfile::supports_hole_punching()` is true. Looks like we need to use `not-generate-blocks-log` as we want this option is available in any platforms.
